### PR TITLE
OCPBUGS-45153: address malformed configmap post-test

### PR DESCRIPTION
**Pull request by rexagod**
- [[Prow]](https://prow.ci.openshift.org/view/gs/test-platform-results/logs/openshift-origin-29307-nightly-4.18-e2e-metal-ipi-ovn-dualstack-techpreview/1858916202634547200) After merging https://github.com/openshift/origin/pull/28889, we found out that the tests were failing on bare-metal, due to the absence of a configmap, and the fact that the tests were "leaving" a new configmap in a malformed state (without a `config.yaml` key).
```
{  fail [github.com/openshift/origin/test/extended/prometheus/prometheus.go:405]: could not determine if Telemetry is enabled: openshift-monitoring/cluster-monitoring-config data lacks a config.yaml key: map[]
Ginkgo exit error 1: exit with code 1}
```
- To address this, https://github.com/openshift/origin/pull/29309 was merged that seemed to [resolve](https://github.com/openshift/origin/pull/29309#issuecomment-2491074917) the issue on bare-metal. Note that with this patch, the configmap was [deleted successfully](https://github.com/openshift/origin/pull/29309#issuecomment-2491074917) upon tests' completion.
- [[Prow]](https://prow.ci.openshift.org/view/gs/test-platform-results/logs/openshift-origin-29309-ci-4.19-e2e-aws-ovn-techpreview/1861901510435147776) Shortly after that, TRT team found out that this introduced a bug for at-least `aws-ovn-techpreview`, stemming from this [snippet](https://github.com/openshift/origin/blob/master/test/extended/prometheus/collection_profiles.go#L100-L105) owing to which the patch was reverted and merged: https://github.com/openshift/origin/pull/29330. However, that snippet is unchanged in the (original and) revert PR, and thus hints towards it being caused due to multiple specs modifying the configmap at the same time.
```
{  fail [github.com/openshift/origin/test/extended/prometheus/collection_profiles.go:180]: Expected
    <*errors.StatusError | 0xc0065a3360>: 
    Operation cannot be fulfilled on configmaps "cluster-monitoring-config": the object has been modified; please apply your changes to the latest version and try again
```
- [[Prow]](https://pr-payload-tests.ci.openshift.org/runs/ci/ead857a0-adb4-11ef-847d-a7a19d6ffba0-0) However, post-revert, on the latest bare-metal must-gathers (for `ipv6` and `dualstack` so far), the CMO configmap collected after the tests are completed has an unpredictable `collectionProfiles` definition (I ran the tests on both 4.18 and 4.19 today, and it seems to be `minimal` and `full`, respectively). Note that earlier, an absence of this patch would have caused the same issue seen in the first point above on bare-metal, but not anymore.
```
apiVersion: v1
data:
  config.yaml: |
    prometheusK8s:
        collectionProfile: "minimal"
kind: ConfigMap
metadata:
  name: cluster-monitoring-config
  namespace: openshift-monitoring
```
***
### [tl;dr](https://github.com/openshift/origin/pull/29331#issuecomment-2508671844)
- Even though the tests may pass post-revert, there's a lingering `collectionProfiles` definition may cause a reduced metrics throughput in general (if set to `minimal`) causing other tests to fail that depend on them.
- Serializing these tests means slightly lesser plaforms to run them on (no bare-metal, for instance).

### CODE REVIEW

**Comment by rexagod** [comment]
/payload-job periodic-ci-openshift-release-master-nightly-4.19-e2e-metal-ipi-ovn-ipv6-techpreview
/payload-job periodic-ci-openshift-release-master-ci-4.19-e2e-aws-ovn-techpreview-serial
/payload-job periodic-ci-openshift-release-master-nightly-4.19-e2e-vsphere-ovn-techpreview-serial

I believe serializing should address the apparent unpredictability. I'll review these jobs once they are done and make any further changes required.

**Comment by rexagod** [comment]
FYI this branch is not rebased on the latest master (that includes the revert). VSphere [results](https://sippy.dptools.openshift.org/sippy-ng/job_runs/1862336532304629760/openshift-origin-29333-nightly-4.19-e2e-vsphere-ovn-techpreview-serial/intervals?filterText=MetricsCollectionProfiles&intervalFile=e2e-timelines_spyglass_20241129-042721.json&overrideDisplayFlag=0&selectedSources=E2EFailed&selectedSources=APIServerGracefulShutdown&selectedSources=E2EPassed&selectedSources=E2ESkipped&selectedSources=E2EFlaked) show no failures, whereas the other serial job failed during bootstrapping. Need to compare this with at-least two other serial jobs to ensure this is deterministic.

**Comment by rexagod** [comment]
Same results as https://github.com/openshift/origin/pull/29333#issuecomment-2507210970 for `e2e-gcp-ovn-techpreview-serial`. Other jobs failed to run due to bootstrapping errors. The order is uncertain, however, there's [no more overlapping](https://sippy.dptools.openshift.org/sippy-ng/job_runs/1862399472235974656/openshift-origin-29333-ci-4.19-e2e-gcp-ovn-techpreview-serial/intervals?filterText=MetricsCollectionProfiles&intervalFile=e2e-timelines_spyglass_20241129-085638.json&overrideDisplayFlag=0&selectedSources=OperatorAvailable&selectedSources=OperatorProgressing&selectedSources=OperatorDegraded&selectedSources=E2EFailed&selectedSources=APIServerGracefulShutdown&selectedSources=KubeEvent&selectedSources=NodeState&selectedSources=E2EPassed&selectedSources=E2ESkipped&selectedSources=E2EFlaked) between specs during their lifetimes, as compared to just their starting times [earlier](https://sippy.dptools.openshift.org/sippy-ng/job_runs/1859513091201437696/openshift-origin-29309-nightly-4.18-e2e-metal-ipi-ovn-ipv6-techpreview/intervals?filterText=MetricsCollectionProfiles&intervalFile=e2e-timelines_spyglass_20241121-105738.json&overrideDisplayFlag=0&selectedSources=OperatorAvailable&selectedSources=OperatorProgressing&selectedSources=OperatorDegraded&selectedSources=E2EFailed&selectedSources=APIServerGracefulShutdown&selectedSources=KubeEvent&selectedSources=NodeState&selectedSources=E2EFlaked&selectedSources=E2EPassed).

The specs themselves may be run in a random order without any conflicts as each innermost `It` processes the configmap to its requirement, but since they don't overlap anymore, this should make things more deterministic as only one spec will ever modify the configmap at any point of time.

**Comment by rexagod** [comment]
/payload-job periodic-ci-openshift-release-master-nightly-4.19-e2e-metal-ipi-ovn-ipv6-techpreview
/payload-job periodic-ci-openshift-multiarch-master-nightly-4.19-ocp-e2e-aws-ovn-techpreview-serial-multi-a-a

**Comment by rexagod** [comment]
### tl;dr

I believe the we need to drop the revert as the patch introduced crucial handling in cases where CMO's configmap wasn't present, i.e., bare-metal. Additionally, we did see reverting that patch fixed the runs on AWS, which (PTAL below) happens owing to a parallelized execution of specs modifying the configmap at the [same time](https://sippy.dptools.openshift.org/sippy-ng/job_runs/1862510185075970048/openshift-origin-29309-nightly-4.19-e2e-gcp-ovn-crun-techpreview/intervals?filterText=MetricsCollectionProfiles&intervalFile=e2e-timelines_spyglass_20241129-162318.json&overrideDisplayFlag=0&selectedSources=OperatorAvailable&selectedSources=OperatorProgressing&selectedSources=OperatorDegraded&selectedSources=KubeletLog&selectedSources=EtcdLog&selectedSources=EtcdLeadership&selectedSources=Alert&selectedSources=Disruption&selectedSources=E2EFailed&selectedSources=APIServerGracefulShutdown&selectedSources=KubeEvent&selectedSources=NodeState). The serialized workflows, to the best of my understanding, should prevent this (as seen from the runs above).

**Comment by rexagod** [comment]
/payload-job periodic-ci-openshift-release-master-ci-4.19-e2e-aws-ovn-techpreview-serial

**Comment by rexagod** [comment]
/payload-job periodic-ci-openshift-release-master-ci-4.19-e2e-aws-ovn-techpreview
/payload-job periodic-ci-openshift-release-master-ci-4.19-e2e-aws-ovn-techpreview-serial

Split up the commit, and explicitly rebased over latest `master` for better visibility.

@sosiouxme @xueqzhan As [stated](https://github.com/openshift/origin/pull/29330#issue-2700034290) in the revert PR, I've applied a fix over the revert commit, and triggered the specified job, as well as it's serialized counterpart, since the tests will run on the latter now.

**Comment by rexagod** [comment]
/payload-job periodic-ci-openshift-release-master-ci-4.19-e2e-aws-ovn-techpreview
/payload-job periodic-ci-openshift-release-master-ci-4.19-e2e-aws-ovn-techpreview-serial

Had to cut the timeout in half, since we don't need to wait that long under serialized conditions (test runs above show a maximum of 4-5m taken by the longest spec).

**Comment by rexagod** [comment]
/payload-job periodic-ci-openshift-release-master-ci-4.19-e2e-aws-ovn-techpreview
/payload-job periodic-ci-openshift-release-master-ci-4.19-e2e-aws-ovn-techpreview-serial

**Comment by rexagod** [comment]
/payload-job periodic-ci-openshift-release-master-ci-4.19-e2e-aws-ovn-techpreview-serial

Jobs keep failing at build time.

**Comment by rexagod** [comment]
/payload-job periodic-ci-openshift-release-master-ci-4.19-e2e-aws-ovn-techpreview-serial

**Comment by rexagod** [comment]
/payload-job periodic-ci-openshift-release-master-ci-4.19-e2e-aws-ovn-techpreview-serial

**Comment by rexagod** [comment]
/payload-job periodic-ci-openshift-release-master-ci-4.19-e2e-gcp-ovn-techpreview-serial

**Comment by rexagod** [comment]
/payload-job periodic-ci-openshift-release-master-ci-4.19-e2e-aws-ovn-techpreview-serial

**Comment by rexagod** [comment]
/payload-job periodic-ci-openshift-release-master-ci-4.19-e2e-gcp-ovn-techpreview-serial
/payload-job periodic-ci-openshift-release-master-ci-4.19-e2e-aws-ovn-techpreview-serial

**Comment by rexagod** [comment]
/payload-job periodic-ci-openshift-release-master-ci-4.19-e2e-gcp-ovn-techpreview-serial
/payload-job periodic-ci-openshift-release-master-ci-4.19-e2e-aws-ovn-techpreview-serial

**Comment by neisw** [comment]
/payload-job periodic-ci-openshift-release-master-ci-4.19-e2e-gcp-ovn-techpreview-serial
/payload-job periodic-ci-openshift-release-master-ci-4.19-e2e-aws-ovn-techpreview-serial

**Comment by rexagod** [comment]
/payload-job periodic-ci-openshift-release-master-ci-4.19-e2e-gcp-ovn-techpreview-serial
/payload-job periodic-ci-openshift-release-master-ci-4.19-e2e-aws-ovn-techpreview-serial

**Comment by rexagod** [comment]
/payload-job periodic-ci-openshift-release-master-ci-4.19-e2e-gcp-ovn-techpreview-serial
/payload-job periodic-ci-openshift-release-master-ci-4.19-e2e-aws-ovn-techpreview-serial

**Comment by rexagod** [comment]
/payload-job periodic-ci-openshift-release-master-ci-4.19-e2e-gcp-ovn-techpreview-serial
/payload-job periodic-ci-openshift-release-master-ci-4.19-e2e-aws-ovn-techpreview-serial

**Comment by rexagod** [comment]
@neisw JUnit seems to report no failing test on AWS, however, the [overall run](https://prow.ci.openshift.org/view/gs/test-platform-results/logs/openshift-origin-29333-ci-4.19-e2e-aws-ovn-techpreview-serial/1864506782295003136) seems to report `context cancelled` and fails after 4h11m.

I see the following in the `build-log`:
```
Step e2e-aws-ovn-techpreview-serial-observers-resource-watch failed after 3h15m14s. 
```

On digging a bit deeper into the aforementioned step:
```
E1205 07:12:59.991552    1522 reflector.go:158] "Unhandled Error" err="k8s.io/client-go@v0.31.1/tools/cache/reflector.go:243: Failed to watch operator.openshift.io/v1, Resource=imagecontentsourcepolicies: failed to list operator.openshift.io/v1, Resource=imagecontentsourcepolicies: the server could not find the requested resource"
```
and,
```
time="2024-12-05T04:26:22Z" level=error msg="disruption sample failed: error running request: 404 Not Found: 404 page not found\n" auditID=23b5d150-89be-460d-88da-4962d9147e2a backend=metrics-api-new-connections this-instance="{Disruption map[backend-disruption-name:metrics-api-new-connections connection:new disruption:openshift-tests]}" type=new
```

**Comment by neisw** [comment]
@rexagod - Yeah looks like they still timed out but you got good test results this time.  The serial job keeps running longer each new test that gets added.  Looks like yours are adding 4 ~ 5 minutes but it's good to see them passing now.

**Comment by rexagod** [comment]
cc @sosiouxme @xueqzhan 

The tests seems okay now, PTAL, thanks!

**Comment by neisw** [comment]
/hold

Let me put up a pr to increase those timeouts.  Don't want that job to start failing if the extra ~5 minutes puts it over the limit.

**Comment by neisw** [comment]
/payload-job periodic-ci-openshift-release-master-ci-4.19-e2e-gcp-ovn-techpreview-serial
/payload-job periodic-ci-openshift-release-master-ci-4.19-e2e-aws-ovn-techpreview-serial

**Comment by rexagod** [comment]
/cc @jan--f

This is good to go now, PTAL 🙂

**Comment by jan--f** [comment]
/lgtm
/approve
Thanks guys!

File: test/extended/prometheus/collection_profiles.go [hunk 1]
```diff
@@ -97,17 +97,31 @@ var _ = g.Describe("[sig-instrumentation][OCPFeatureGate:MetricsCollectionProfil
...
 			g.By("restoring the original configuration for the operator")
 			_, err = r.kclient.CoreV1().ConfigMaps(operatorNamespaceName).Update(tctx, currentConfiguration, metav1.UpdateOptions{})
 		} else {
+			shouldDeleteConfiguration = true
 			g.By("cleaning up the configuration for the operator as it did not exist pre-job")
 			err = r.kclient.CoreV1().ConfigMaps(operatorNamespaceName).Delete(tctx, operatorConfigurationName, metav1.DeleteOptions{})
 		}
 		o.Expect(err).To(o.BeNil())
+
+		o.Eventually(func() error {
```
Comment (machine424) [inline test/extended/prometheus/collection_profiles.go:114]: As I'm not a big user of `gomega`, I'm interested to know why this double check is needed, I'd expected the deletion +  `o.Expect(err).To(o.BeNil())` above to be sufficient, if not, it's a bug in `Delete` or `gomega`

Comment (rexagod) [reply test/extended/prometheus/collection_profiles.go:114]: I wanted to ensure here that the configmap is actually deleted before the test exits.

**Comment by rexagod** [comment]
Thanks, Jan. It seems this also needs a critical fix and an approved label.

File: test/extended/prometheus/collection_profiles.go (hunk 1 above)
Comment (machine424) [reply test/extended/prometheus/collection_profiles.go:114]: AFAIK, the default invocation of `Delete()` uses foreground deletion which waits for the resource to disappear before returning.
But maybe I'm missing sth.
It's just a remark though, not trying to block the merge.

Comment (machine424) [reply test/extended/prometheus/collection_profiles.go:114]: Seeing https://github.com/kubernetes/kubernetes/blob/9dd53b55f7b21deb8d7cc4f096899af8f46bd083/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L551 made me think that foreground is the default.
You're right, it defaults to background.

So maybe, instead of Getting the configmap and fail the test if it's still there (for some reason, an overwhelmed api or sth who knows), maybe we can just tell the api to use foreground via the `DeleteOptions`, in that case we give the deletion more chance/time.

**Review by sosiouxme — APPROVED** [review]
/approve

**Comment by rexagod** [comment]
/hold cancel

Unrelated node unit tests were flaking, but passing now, so removing the [hold](https://github.com/openshift/origin/pull/29333#issuecomment-2574296725).

**Comment by jan--f** [comment]
/lgtm
But there seems to be a merge conflict now?

File: test/extended/prometheus/collection_profiles.go [hunk 2]
```diff
@@ -120,7 +121,7 @@ var _ = g.Describe("[sig-instrumentation][OCPFeatureGate:MetricsCollectionProfil
 			}
 
 			return nil
-		}).Should(o.BeNil())
+		}, pollTimeout, pollInterval).Should(o.BeNil())
```
Comment (rexagod) [inline test/extended/prometheus/collection_profiles.go:124]: Owing to the changes introduced in https://github.com/openshift/origin/pull/29428. 

Please `/lgtm` again, thanks.

**Comment by rexagod** [comment]
Serial tests may be overridden as the failures are unrelated: https://github.com/openshift/origin/pull/29459#issuecomment-2608723379.

**Comment by rexagod** [comment]
Rebased on `master` to include the recent changes from https://github.com/openshift/origin/pull/29459 (and hopefully get the tests green, although they overrode one of the flaky serial tests so we might have to do that same).

**Comment by rexagod** [comment]
/payload-job periodic-ci-openshift-release-master-ci-4.19-e2e-gcp-ovn-techpreview-serial
/payload-job periodic-ci-openshift-release-master-ci-4.19-e2e-aws-ovn-techpreview-serial

**Comment by rexagod** [comment]
cc @sosiouxme @jan--f for re-LGTM here, thanks!
